File: frontend/davon-library-webui/src/lib/api/services/favorite.service.ts
import { httpClient } from '../client';

export interface Favorite {
  id: number;
  book: {
    id: number;
    title: string;
    author: {
      firstName: string;
      lastName: string;
    };
    category: {
      name: string;
    };
  };
  createdAt: string;
}

export const favoriteService = {
  /**
   * Get all favorites for a user
   */
  getUserFavorites: async (userId: number): Promise<Favorite[]> => {
    return await httpClient.get(`/api/favorites/user/${userId}`);
  },

  /**
   * Add a book to user's favorites
   */
  addFavorite: async (userId: number, bookId: number): Promise<Favorite> => {
    return await httpClient.post(`/api/favorites/user/${userId}/book/${bookId}`);
  },

  /**
   * Remove a book from user's favorites
   */
  removeFavorite: async (userId: number, bookId: number): Promise<void> => {
    await httpClient.delete(`/api/favorites/user/${userId}/book/${bookId}`);
  },

  /**
   * Toggle favorite status
   */
  toggleFavorite: async (userId: number, bookId: number): Promise<{ isFavorited: boolean; message: string }> => {
    return await httpClient.post(`/api/favorites/user/${userId}/book/${bookId}/toggle`);
  },

  /**
   * Check if a book is favorited by a user
   */
  getFavoriteStatus: async (userId: number, bookId: number): Promise<{ isFavorited: boolean }> => {
    return await httpClient.get(`/api/favorites/user/${userId}/book/${bookId}/status`);
  },

  /**
   * Get favorite count for a book
   */
  getFavoriteCount: async (bookId: number): Promise<{ count: number }> => {
    return await httpClient.get(`/api/favorites/book/${bookId}/count`);
  }
};

File: backend/davon-library-backend/src/main/java/org/acme/service/FavoriteService.java
package org.acme.service;

import org.acme.model.Favorite;
import org.acme.model.Book;
import org.acme.model.User;
import org.acme.repository.FavoriteRepository;
import org.acme.repository.BookRepository;
import org.acme.repository.UserRepository;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import java.util.List;
import java.util.Optional;

@ApplicationScoped
public class FavoriteService {

    @Inject
    private FavoriteRepository favoriteRepository;

    @Inject
    private BookRepository bookRepository;

    @Inject
    private UserRepository userRepository;

    /**
     * Get all favorites for a user
     */
    @Transactional
    public List<Favorite> getUserFavorites(Long userId) {
        List<Favorite> favorites = favoriteRepository.findByUserId(userId);

        // Force loading of related entities
        for (Favorite favorite : favorites) {
            if (favorite.getBook() != null) {
                Book book = favorite.getBook();
                book.getTitle();
                if (book.getAuthor() != null) {
                    book.getAuthor().getFirstName();
                    book.getAuthor().getLastName();
                }
                if (book.getCategory() != null) {
                    book.getCategory().getName();
                }
            }
        }

        return favorites;
    }

    /**
     * Add a book to user's favorites
     */
    @Transactional
    public Favorite addFavorite(Long userId, Long bookId) {
        // Check if already favorited
        Optional<Favorite> existing = favoriteRepository.findByUserIdAndBookId(userId, bookId);
        if (existing.isPresent()) {
            return existing.get();
        }

        // Validate user and book exist
        Optional<User> user = userRepository.findByIdOptional(userId);
        Optional<Book> book = bookRepository.findByIdOptional(bookId);

        if (user.isEmpty()) {
            throw new IllegalArgumentException("User not found with id: " + userId);
        }
        if (book.isEmpty()) {
            throw new IllegalArgumentException("Book not found with id: " + bookId);
        }

        // Create new favorite
        Favorite favorite = new Favorite();
        favorite.setUser(user.get());
        favorite.setBook(book.get());

        favoriteRepository.persist(favorite);
        return favorite;
    }

    /**
     * Remove a book from user's favorites
     */
    @Transactional
    public boolean removeFavorite(Long userId, Long bookId) {
        return favoriteRepository.deleteByUserIdAndBookId(userId, bookId);
    }

    /**
     * Check if a book is favorited by a user
     */
    public boolean isFavorited(Long userId, Long bookId) {
        return favoriteRepository.isFavorited(userId, bookId);
    }

    /**
     * Get favorite count for a book
     */
    public long getFavoriteCount(Long bookId) {
        return favoriteRepository.getFavoriteCountForBook(bookId);
    }

    /**
     * Toggle favorite status
     */
    @Transactional
    public boolean toggleFavorite(Long userId, Long bookId) {
        if (isFavorited(userId, bookId)) {
            removeFavorite(userId, bookId);
            return false; // Removed from favorites
        } else {
            addFavorite(userId, bookId);
            return true; // Added to favorites
        }
    }
}